### PR TITLE
Optimize substitute method

### PR DESCRIPTION
**Pull request by jalalium**
### Description
This PR addresses an optimization for the substitute method in the PyccelAstNode class. The current implementation does not check if original and replacement are empty, leading to unnecessary recursion over the code. This update adds a condition to return immediately if  original is empty (and consequenty replacement), thereby improving the performance of the method.

### CODE REVIEW

File: .dict_custom.txt [hunk 1]
```diff
@@ -119,3 +119,4 @@ indexable
 traceback
 STC
 gFTL
+Optimize
```
Comment (EmilyBourne) [inline .dict_custom.txt:122]: optimize should be in the dictionary. Probably it is missing because we are using UK english

File: CHANGELOG.md [hunk 2]
```diff
@@ -27,6 +27,7 @@ All notable changes to this project will be documented in this file.
 -   #1895 : Add Python support for dict initialisation with `dict()`.
 -   #1886 : Add Python support for dict method `pop()`.
 -   #1936 : Add missing C output for inline decorator example in documentation
+-   #1937 : Optimize substitute method
```
Comment (EmilyBourne) [inline CHANGELOG.md:30]:
```suggestion
-   #1937 : Optimise `pyccel.ast.basic.PyccelAstNode.substitute` method.
```

File: .dict_custom.txt (hunk 1 above)
Comment (jalalium) [reply .dict_custom.txt:122]: Shall I remove it and use `optimise` instead?

Comment (EmilyBourne) [reply .dict_custom.txt:122]: Yes please

Comment (jalalium) [reply .dict_custom.txt]: done

**Review by EmilyBourne — APPROVED** [review]

**Pull request merged**